src/components/MediaPlayer/VideoJS/VideoJSPlayer.js

```diff
@@ -56,293 +52,171 @@ function VideoJSPlayer({
     canvasLink,
-    currentNavItem,
     hasMultiItems,
-    srcIndex,
     targets,
     autoAdvance,
-    playlist,
     structures,
     canvasSegments,
     hasStructure,
-    canvasIsEmpty,
   } = manifestState;
-  const {
-    isClicked,
-    isEnded,
-    isPlaying,
-    player,
-    searchMarkers,
-    currentTime,
-  } = playerState;
-
-  const [cIndex, _setCIndex] = React.useState(canvasIndex);
-  const [isReady, _setIsReady] = React.useState(false);
-  const [activeId, _setActiveId] = React.useState('');
+  const { isEnded, isPlaying, currentTime } = playerState;
+
   const [startVolume, setStartVolume] = useLocalStorage('startVolume', 1);
-  const [startQuality, setStartQuality] = useLocalStorage('startQuality', null);
   const [startMuted, setStartMuted] = useLocalStorage('startMuted', false);
   const [startCaptioned, setStartCaptioned] = useLocalStorage('startCaptioned', true);
-  const [fragmentMarker, setFragmentMarker] = React.useState(null);
-  const [messageTime, setMessageTime] = React.useState(CANVAS_MESSAGE_TIMEOUT / 1000);
-
-  const videoJSRef = React.useRef(null);
-  const playerRef = React.useRef(null);
-
-  const autoAdvanceRef = React.useRef();
-  autoAdvanceRef.current = autoAdvance;
-
-  const srcIndexRef = React.useRef();
-  srcIndexRef.current = srcIndex;
-
-  let activeIdRef = React.useRef();
-  activeIdRef.current = activeId;
-  const setActiveId = (id) => {
-    _setActiveId(id);
-    activeIdRef.current = id;
-  };
-
-  let currentTimeRef = React.useRef();
-  currentTimeRef.current = currentTime;
-
-  let isReadyRef = React.useRef();
-  isReadyRef.current = isReady;
-  const setIsReady = (r) => {
-    _setIsReady(r);
-    isReadyRef.current = r;
-  };
-
-  let currentNavItemRef = React.useRef();
-  currentNavItemRef.current = currentNavItem;
-
-  let canvasIsEmptyRef = React.useRef();
-  canvasIsEmptyRef.current = canvasIsEmpty;
-
-  let canvasDurationRef = React.useRef();
-  canvasDurationRef.current = canvasDuration;
-
-  let canvasLinkRef = React.useRef();
-  canvasLinkRef.current = canvasLink;
+  const [startQuality, setStartQuality] = useLocalStorage('startQuality', null);
 
-  let isPlayingRef = React.useRef();
-  isPlayingRef.current = isPlaying;
+  const videoJSRef = useRef(null);
+  const captionsOnRef = useRef();
+  const activeTrackRef = useRef();
 
-  let isEndedRef = React.useRef();
-  isEndedRef.current = isEnded;
+  const { canvasIndex, canvasIsEmpty, lastCanvasIndex } = useMediaPlayer();
+  const { isPlaylist, renderingFiles, srcIndex, switchPlayer }
+    = useSetupPlayer({ enableFileDownload, withCredentials, lastCanvasIndex });
+  const { messageTime } = useShowInaccessibleMessage({ lastCanvasIndex });
 
-  let cIndexRef = React.useRef();
-  cIndexRef.current = canvasIndex;
-  const setCIndex = (i) => {
-    _setCIndex(i);
-    cIndexRef.current = i;
-  };
+  const canvasIsEmptyRef = useRef();
+  canvasIsEmptyRef.current = useMemo(() => { return canvasIsEmpty; }, [canvasIsEmpty]);
```

Comment:
This is a neat pattern for updating a ref using hooks.  How does this compare with what react would do with useState and not a ref?

Reply:
There are couple of advantages to using refs in place of state;
- Using refs in place of state helps to control re-renders of the component as state variables automatically triggers a component re-render
- In this particular instance, when using these variables inside player's event handlers it seems variables defined as refs carry the most recently updated value rather than the variables defined as state.
- Using `useMemo` to update the refs ensures that the value of the variable is persisted across re-renders without corruption, since it is only updated when the dependency of `useMemo` is changed.